keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -326,75 +287,114 @@ spec:
 
+We will document this requirement in our documentation that faces CSI driver vendors in gihub.com/kubernetes-csi/docs.
 
+### Risks and Mitigations
 
-### Implementation Details/Notes/Constraints [optional]
+This KEP changes behavior of Kubernetes when two pods with different SELinux contexts use the same volume. See [Behavioral changes](#behavioral-changes) above.
+There is a risk that existing applications will get broken when the feature is enabled.
+To mitigate this risk, we propose:
 
-### Risks and Mitigations
+* Implement this feature in [several phases](#Implementation-phases).
+* Expose metrics + alerts before the feature is enabled by default and could break anyone.
 
 ## Design Details
 
+### Required kubelet changes
+
+Apart from the obvious API change and behavior described above, kubelet + volume plugins need not so obvious changes: 
+
+* Kubelet's VolumeManager needs to track which SELinux label should get a volume in global mount (to call `MountDevice()` with the right mount options).
+  * It must call `UnmountDevice()` even when another pod wants to re-use a mounted volume, but it has a different SELinux context.
+  * After kubelet restart, kubelet must reconstruct the original SELinux label it used to SetUp and MountDevice of each volume.
+    * Volume reconstruction must be updated to get the SELinux label from mount (in-tree volume plugins) or stored json file (CSI).
+      This label must be updated in VolumeManager's ActualStateOfWorld after reconstruction.
+  * Reconciler must check also SELinux context used to mount a volume (both mounted devices and volumes) before considering what operation to take on a volume (`MountVolume` or `UnmountVolume`/`UnmountDevice` or nothing).
+    It must throw proper error message telling that a Pod can't start because its volume is used by another Pod with a different SELinux context.
+    * This is a good point to capture any metrics proposed below.
+* Volume plugins will get SELinux context as a new parameter of `MountDevice` and `SetUp`/`SetupAt` calls (resp. as a new field in `DeviceMounterArgs` / `MounterArgs`).
+  * Each volume plugin can choose to use the mount option `-o context=` (e.g. when `CSIDriver.SELinuxRelabelPolicy` is `true`) or ignore it (e.g. in-tree volume plugins for shared filesystems or when `CSIDriver.SELinuxRelabelPolicy` is `false` or `nil`).
+  * Each volume plugin then returns `SupportsSELinux` from `GetAttributes()` call, depending on if it wants the container runtime to relabel the volume (`true`) or not (`false`; the volume was already mounted with the right label or it does not support SELinux at all).
+* When a CSI driver announces `SELinuxMountSupported: true`, kubelet will check that `-o context=X` was correctly applied after `NodePublish()`.
+  It will report error when the context in `/proc/mounts` does not match the expected value.
```

Comment:
My initial idea that it will be a `SetUp()` error, i.e. kubelet will retry `SetUp()` again, until the driver fixes it. The pod won't start.

Question is, if the driver can fix it, if it mounted the volume wrong at the first attempt - any retry probably won't help much here.

Reply:
Rewording the bullet to mention the consequence - the pod won't ever start, the CSI driver is faulty.